spec/term/1.0.md

```diff
@@ -0,0 +1,116 @@
+
+Finally, the server MUST discard each U+000A (new line) character that immediately succeeds a U+000D (carriage return) character, and then convert each U+000D character into a U+000A character.
+
+*Rationale:* This rule ensures consistent line break handling across different platforms, i.e., CR+NL = CR = NL.
+
+The remaining string of Unicode characters SHALL then be written into the character array starting from the position pointed to by the cursor, using the rules for normalization, writing direction, line breaking and handling of bidirectional text as laid out in the Unicode Standard, as published by the [Unicode consortium](https://www.unicode.org).
+Combining characters are written into the same position as their base characters.
+
+The cursor SHALL be updated to point to that position in the character array where, if the next character to be written to standard output by a client was U+FFFD (Replacement Character), that character would be written.
+
+*Rationale:* The Replacement Character is chosen because it is particularly boring. It has a neutral directionality, does not combine with anything, and does not change change line-breaking or hyphenation rules.
+
+### 1.3. Standard input
+
+TODO describe how keypresses are reflected on stdin (canonical/raw mode, noecho property, ), how Ctrl-D is handled
+
+## 2. Module dependencies
+
+This module DEPENDS ON any version of the [`vt6/core` module](https://vt6.io/std/core/).
+
+## 3. Properties for `vt6/term1`
+
+### 3.1. The `term1.width` property
+
+TODO unsigned integer (but not zero)
+
+### 3.2. The `term1.viewport-height` property
+
+TODO unsigned integer (zero means infinitely tall viewport, e.g. continous stationery)
+
+### 3.3. The `term1.scrollback-height` property
+
+TODO unsigned integer (zero means infinitely many scrollback, e.g. on GUI terminal emulator or continous stationery)
+
+### 3.4. The `term1.input-raw` property
+
+TODO disabled by default; when enabled, server MUST make all keyboard input available on stdin immediately instead of waiting for a full line
+
+### 3.5. The `term1.input-echo` property
+
+TODO enabled by default; when enabled, printable characters entered by the user are reflected on the terminal; when disabled, may only be re-enabled on the server connection where it was disabled; will be re-enabled automatically when the server connection disabling it is closed
+
+### 3.6. The `term1.output-protected` property
+
+TODO disabled by default; when enabled, server MUST disregard all unprintable characters and control characters (incl. \<ESC\>) except for CR/LF/HT on any standard input; may only be disabled on the server connection where it was enabled; will be disabled automatically when the server connection enabling it is closed (the purpose of this is to prevent exploits where carefully crafted shell scripts appear benign when printed with `cat(1)`, but contain malicious code which is hidden with backspace chars and VT100 control sequences)
```

Comment:
> may be disabled only on othe server connection where it was enabled

This creates a race: When stdout ends with a control sequence, whether it is interpreted depends on whether the terminal drains stdout first, or reads and executes the `(core1.set term1.output-protected false)` instruction first.

I think it's inevitable that we come up with some sort of equivalent to `TCSA{NOW,DRAIN,FLUSH}` (see `man 3 tcsetattr`).

Reply:
Correction: There is no race. `core1.set` is answered with `core1.pub`, which the application can wait for before starting to write the protected output data.